pennylane/transforms/decompose.py

```diff
@@ -0,0 +1,229 @@
+"""
+A transform for decomposing quantum circuits into user defined gate sets. Offers an alternative to the more device-focused decompose transform.
+"""
+# pylint: disable=protected-access
+# pylint: disable=unnecessary-lambda-assignment
+
+import warnings
+from collections.abc import Callable, Generator, Sequence
+from typing import Iterable, Optional
+
+import pennylane as qml
+from pennylane.transforms.core import transform
+
+
+def null_postprocessing(results):
+    """A postprocessing function returned by a transform that only converts the batch of results
+    into a result for a single ``QuantumTape``.
+    """
+    return results[0]
+
+
+def _operator_decomposition_gen(
+    op: qml.operation.Operator,
+    acceptance_function: Callable[[qml.operation.Operator], bool],
+    decomposer: Callable[[qml.operation.Operator], Sequence[qml.operation.Operator]],
+    max_expansion: Optional[int] = None,
+    current_depth=0,
+) -> Generator[qml.operation.Operator, None, None]:
+    """A generator that yields the next operation that is accepted."""
+
+    max_depth_reached = False
+
+    if max_expansion is not None and max_expansion <= current_depth:
+        max_depth_reached = True
+
+    if acceptance_function(op) or max_depth_reached:
+        yield op
+    else:
+        decomp = decomposer(op)
+        current_depth += 1
+
+        for sub_op in decomp:
+            yield from _operator_decomposition_gen(
+                sub_op,
+                acceptance_function,
+                decomposer=decomposer,
+                max_expansion=max_expansion,
+                current_depth=current_depth,
+            )
+
+
+@transform
+def decompose(tape, gate_set=None, max_expansion=None):
+    """Decomposes a quantum circuit into a user-specified gate set.
+
+    Args:
+        tape (QuantumScript or QNode or Callable): a quantum circuit.
+        gate_set (Iterable[Union[str, type]] or Callable[Operator, bool], optional): Decomposition gates defined by either (1) a gate set of operators or (2) a rule that they must follow.
+            Defaults to None. If ``None``, gate set defaults to all available operators given by ``~.pennylane.ops.__all__``.
+        max_expansion (int, optional): The maximum depth of the decomposition. Defaults to None. If ``None``, circuit will be decomposed until the target gate set is reached.
+
+    Returns:
+        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumScript], function]:
+
+        The decomposed circuit. The output type is explained in :func:`qml.transform <pennylane.transform>`.
+
+    .. seealso:: :func:`~.pennylane.devices.preprocess.decompose` for a transform that is intended for device developers. This function decomposes a quantum circuit into a set of basis gates available on a given device architecture.
+
+    **Example**
+
+    Consider the following tape:
+
+    >>> tape = qml.tape.QuantumScript([qml.IsingXX(1.2, wires=(0,1))], [qml.expval(qml.Z(0))])
```

Comment:
If you aren't full-screen on a large monitor, this line is a little long, and creates a scrollbar. Splitting it up will avoid that.

```suggestion
    >>> ops = [qml.IsingXX(1.2, wires=(0,1))]
    >>> tape = qml.tape.QuantumScript(ops, [qml.expval(qml.Z(0))])
```